File: sorting/insertion_sort.rb
# Keeps a sorted and unsorted portin of array. Starting with the first value, compares value to the left and shifts if the vlaue is larger. Otherwise, value will assume to be sorted and teh process is repeated for the remainder of the array.

def insertionSort(input)
  input.each_with_index do |num, i|
    temp = input[i]
    pointer = i
    # shift value down to left until smaller value is found
    while pointer > 0 && temp < input[pointer - 1]
      input[pointer] = input[pointer - 1]
      pointer -= 1
    end
    input[pointer] = temp
  end
  input
end


require 'test/unit'

class InsertionSortTest < Test::Unit::TestCase
  def test_insertionSort_should_handle_example_case
    test = insertionSort([3,9,1,4,7])
    answer = [1,3,4,7,9]

    assert_equal(answer, test);
  end

  def test_insertionSort_should_handle_empty_input
    test = insertionSort([])
    answer = []

    assert_equal(answer, test);
  end

  def test_insertionSort_should_handle_large_input
    testInput = []
    $i = 1000000

    while $i > 0
      toPush = Random.rand(1000000)
      testInput.push(toPush)
      $i = $i - 1
    end

    test = insertionSort(testInput)
    assert_equal(testInput.sort, test)
  end

end

File: sorting/insertion_sort.js
var insertionSort = function(input){
  for (var i = 0; i < input.length; i++) {
    var temp = input[i];
    var pointer = i;
    while (pointer > 0 && temp < input[pointer - 1]) {
      input[pointer] = input[pointer - 1];
      pointer--;
    }
    input[pointer] = temp;
  }
  return input;
}

// var expect = require('chai').expect;

// describe('INSERTION SORT ', function(){

//   describe("RUN ON [8, 3, 2, 10]: ", function(){
//     it("should return [2, 3, 8, 10]", function(){
//       var test = insertionSort([3,9,1,4,7]);
//       var answer = [1,3,4,7,9];

//       expect(test).to.eql(answer);
//     });
//   });

//   describe("RUN ON EMPTY INPUT []: ", function(){
//     it("should return []", function(){
//       var test = insertionSort([]);
//       var answer = [];

//       expect(test).to.eql(answer);
//     });
//   });

//   describe("RUN ON LARGE INPUT: ", function(){
//     it("should complete", function(){
//       var testInput = [];
//       var i = 1000000;
//       while (i--){
//         testInput.push(Math.floor(Math.random() * 1000000))
//       }
//       var test = insertionSort(testInput);

//       expect(testInput.sort(function(a, b){return a-b})).to.eql(test);
//     })
//   })

// });
